### ISSUE TITLE
Create `distillation/trainer.py`

### ISSUE THREAD

**Issue by peremartra** [peremartra/optipfair#24]
# Sub-issue 3/5: Create `distillation/trainer.py`

**Parent issue:** #21 
**Depends on:** Sub-issue 1 (mapping) and Sub-issue 2 (loss)

---

## Context

This task implements `distill_model()`, the main public API of the distillation module. It is an adaptation of `train_student_advanced()` from `CH06_NB03_Hands_on_qwen3_5.ipynb`, integrated into OptiPFair conventions.

The function manages the full training loop including gradient accumulation, progress tracking, and stats reporting. It is consistent with the existing `prune_model()` API: receives models already loaded in memory, and returns the trained model or a `(model, stats_dict)` tuple when `return_stats=True`.

---

## Files to Create
```
optipfair/distillation/trainer.py       # the implementation
tests/test_distillation_trainer.py      # tests
```

---

## Reference Implementation

The following is the original training loop from `CH06_NB03_Hands_on_qwen3_5.ipynb`:
```python
def train_student_advanced(
    student_model,
    teacher_model,
    dataloader,
    layer_map=None,
    alpha=0.1, beta=0.8, gamma=0.05, delta=0.05,
    temperature=2.0, skew_alpha=0.5,
    epochs=3, learning_rate=4e-5,
    experiment_name="experiment", accumulation_steps=4
):
    optimizer = torch.optim.AdamW(student_model.parameters(), lr=learning_rate)
    student_model.train()
    teacher_model.eval()

    if layer_map is None:
        num_layers = student_model.config.num_hidden_layers
        layer_map = list(range(num_layers))

    request_hidden_states = (gamma > 0 or delta > 0)
    loss_history = {'total': [], 'task': [], 'logits': [], 'trajectory': [], 'derivative': []}
    epoch_times = []
    total_start_time = time.time()

    for epoch in range(epochs):
        epoch_start_time = time.time()
        epoch_losses = {k: [] for k in loss_history.keys()}
        progress_bar = tqdm(dataloader, desc=f"Epoch {epoch+1}/{epochs}")
        accumulated_losses = {k: 0.0 for k in loss_history.keys()}
        accumulation_counter = 0

        for batch_idx, (input_ids, attention_mask) in enumerate(progress_bar):
            input_ids = input_ids.to(device)
            attention_mask = attention_mask.to(device)
            labels = input_ids.clone()

            student_outputs = student_model(
                input_ids=input_ids, attention_mask=attention_mask,
                output_hidden_states=request_hidden_states
            )
            with torch.no_grad():
                teacher_outputs = teacher_model(
                    input_ids=input_ids, attention_mask=attention_mask,
                    output_hidden_states=request_hidden_states
                )

            student_hiddens = (student_outputs.hidden_states[1:]
                               if request_hidden_states else None)
            teacher_hiddens = (teacher_outputs.hidden_states[1:]
                               if request_hidden_states else None)

            loss, loss_dict = compute_compound_loss_advanced(
                student_logits=student_outputs.logits,
                teacher_logits=teacher_outputs.logits,
                student_hiddens=student_hiddens,
                teacher_hiddens=teacher_hiddens,
                labels=labels, layer_map=layer_map,
                alpha=alpha, beta=beta, gamma=gamma, delta=delta,
                temperature=temperature, skew_alpha=skew_alpha
            )

            scaled_loss = loss / accumulation_steps
            scaled_loss.backward()

            for key in accumulated_losses:
                accumulated_losses[key] += loss_dict[key]
            accumulation_counter += 1

            if (batch_idx + 1) % accumulation_steps == 0:
                torch.nn.utils.clip_grad_norm_(student_model.parameters(), max_norm=1.0)
                optimizer.step()
                optimizer.zero_grad()
                avg_losses = {k: v / accumulation_counter for k, v in accumulated_losses.items()}
                for key in avg_losses:
                    epoch_losses[key].append(avg_losses[key])
                accumulated_losses = {k: 0.0 for k in loss_history.keys()}
                accumulation_counter = 0

        if accumulation_counter > 0:
            torch.nn.utils.clip_grad_norm_(student_model.parameters(), max_norm=1.0)
            optimizer.step()
            optimizer.zero_grad()
            avg_losses = {k: v / accumulation_counter for k, v in accumulated_losses.items()}
            for key in avg_losses:
                epoch_losses[key].append(avg_losses[key])

        for key in epoch_losses:
            if epoch_losses[key]:
                loss_history[key].append(np.mean(epoch_losses[key]))

        epoch_time = time.time() - epoch_start_time
        epoch_times.append(epoch_time)

    total_time = time.time() - total_start_time
    loss_history['epoch_times_seconds'] = epoch_times
    loss_history['total_time_seconds'] = total_time
    return student_model, loss_history
```

---

### `optipfair/distillation/trainer.py`

Adapt the reference above as `distill_model()` with:

**Signature:**
```python
def distill_model(
    student_model: PreTrainedModel,
    teacher_model: PreTrainedModel,
    dataloader: DataLoader,
    layer_mapping_strategy: str = "uniform",  # "uniform" | "last"
    alpha: float = 0.6,    # Hard labels weight: learns from ground truth (dataset)
    beta: float = 0.4,     # Soft labels weight: learns from teacher's output distribution
    gamma: float = 0.0,    # Feature alignment weight: matches hidden state representations layer-by-layer
    delta: float = 0.0,    # Feature dynamics weight: matches how representations change between layers
    temperature: float = 2.0,
    skew_alpha: float = 0.4,
    epochs: int = 3,
    learning_rate: float = 4e-5,
    accumulation_steps: int = 4,
    show_progress: bool = True,
    return_stats: bool = False,
) -> Union[PreTrainedModel, Tuple[PreTrainedModel, Dict]]
```

**Key adaptations from the reference:**

1. Replace the `layer_map` parameter with `layer_mapping_strategy`. Call `resolve_layer_map()` from `mapping.py` internally — only when `gamma > 0` or `delta > 0`.

2. Replace `compute_compound_loss_advanced` call with `compute_distillation_loss` from `loss.py`.

3. Infer `device` from the student model (`next(student_model.parameters()).device`) instead of relying on a global variable.

4. When `return_stats=False`, return only `trained_model`. When `return_stats=True`, return `(trained_model, stats_dict)`.

**`stats_dict` structure:**
```python
{
    "epochs":                3,
    "learning_rate":         4e-5,
    "accumulation_steps":    4,
    "effective_batch_size":  64,      # dataloader.batch_size * accumulation_steps
    "alpha":       0.6,
    "beta":        0.4,
    "gamma":       0.0,
    "delta":       0.0,
    "temperature": 2.0,
    "skew_alpha":  0.4,
    "layer_mapping_strategy": "uniform",   # None if gamma=0 and delta=0
    "layer_map_used":         [0, 2, 4],   # None if gamma=0 and delta=0
    "loss_history": {
        "total":        [2.31, 1.87, 1.62],
        "task":         [...],
        "logits":       [...],
        "trajectory":   [...],
        "derivative":   [...],
    },
    "epoch_times_seconds": [138.2, 137.9, 138.5],
    "total_time_seconds":  414.6,
    "avg_time_per_epoch":  138.2,
}
```

**Validations to implement:**
```python
if alpha + beta + gamma + delta == 0:
    raise ValueError("At least one loss weight (alpha, beta, gamma, delta) must be > 0.")

if student_model is teacher_model:
    raise ValueError("student_model and teacher_model must be different objects.")

if layer_mapping_strategy not in (MAPPING_UNIFORM, MAPPING_LAST):
    raise ValueError(f"Unsupported layer_mapping_strategy: '{layer_mapping_strategy}'.")

if (gamma > 0 or delta > 0) and alpha + beta + gamma + delta != 1.0:
    import warnings
    warnings.warn("Loss weights do not sum to 1.0. This is valid but may affect training scale.")
```

---

## Tests

Create `tests/test_distillation_trainer.py` using mock models following the pattern of `tests/test_depth.py`:
```python
"""
Tests for the distillation trainer module.
"""

import unittest
import torch
from torch import nn
from torch.utils.data import DataLoader, TensorDataset
import sys
import os

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from optipfair.distillation.trainer import distill_model
from optipfair.distillation.mapping import MAPPING_UNIFORM, MAPPING_LAST


class MockConfig:
    def __init__(self, num_hidden_layers=4, hidden_size=32):
        self.num_hidden_layers = num_hidden_layers
        self.hidden_size = hidden_size


class MockAttentionOutput:
    def __init__(self, hidden_size):
        self.hidden_size = hidden_size


class MockModelOutput:
    def __init__(self, logits, hidden_states=None):
        self.logits = logits
        self.hidden_states = hidden_states


class MockSmallModel(nn.Module):
    """Minimal mock model that returns logits and optional hidden states."""

    def __init__(self, num_layers=4, hidden_size=32, vocab_size=100):
        super().__init__()
        self.config = MockConfig(num_layers, hidden_size)
        self.embedding = nn.Embedding(vocab_size, hidden_size)
        self.layers = nn.ModuleList([
            nn.Linear(hidden_size, hidden_size) for _ in range(num_layers)
        ])
        self.lm_head = nn.Linear(hidden_size, vocab_size)
        self._num_layers = num_layers
        self._hidden_size = hidden_size

    def forward(self, input_ids, attention_mask=None, output_hidden_states=False):
        x = self.embedding(input_ids)
        hidden_states = []
        for layer in self.layers:
            x = layer(x)
            if output_hidden_states:
                hidden_states.append(x)
        logits = self.lm_head(x)
        hs = tuple([self.embedding(input_ids)] + hidden_states) if output_hidden_states else None
        return MockModelOutput(logits=logits, hidden_states=hs)


def make_dataloader(batch_size=2, seq_len=8, vocab_size=100, num_batches=3):
    input_ids = torch.randint(0, vocab_size, (batch_size * num_batches, seq_len))
    attention_mask = torch.ones_like(input_ids)
    dataset = TensorDataset(input_ids, attention_mask)
    return DataLoader(dataset, batch_size=batch_size)


class TestDistillModelValidation(unittest.TestCase):

    def setUp(self):
        self.student = MockSmallModel(num_layers=4)
        self.teacher = MockSmallModel(num_layers=6)
        self.dataloader = make_dataloader()

    def test_raises_if_all_weights_zero(self):
        with self.assertRaises(ValueError):
            distill_model(
                self.student, self.teacher, self.dataloader,
                alpha=0.0, beta=0.0, gamma=0.0, delta=0.0
            )

    def test_raises_if_same_model_object(self):
        with self.assertRaises(ValueError):
            distill_model(self.student, self.student, self.dataloader)

    def test_raises_if_invalid_strategy(self):
        with self.assertRaises(ValueError):
            distill_model(
                self.student, self.teacher, self.dataloader,
                layer_mapping_strategy="original_indices",
                gamma=0.1
            )


class TestDistillModelReturnValues(unittest.TestCase):

    def setUp(self):
        self.student = MockSmallModel(num_layers=4)
        self.teacher = MockSmallModel(num_layers=6)
        self.dataloader = make_dataloader()

    def test_returns_model_when_return_stats_false(self):
        result = distill_model(
            self.student, self.teacher, self.dataloader,
            epochs=1, show_progress=False, return_stats=False
        )
        self.assertIsInstance(result, nn.Module)

    def test_returns_tuple_when_return_stats_true(self):
        result = distill_model(
            self.student, self.teacher, self.dataloader,
            epochs=1, show_progress=False, return_stats=True
        )
        self.assertIsInstance(result, tuple)
        self.assertEqual(len(result), 2)

    def test_stats_dict_has_required_keys(self):
        _, stats = distill_model(
            self.student, self.teacher, self.dataloader,
            epochs=1, show_progress=False, return_stats=True
        )
        required_keys = {
            "epochs", "learning_rate", "accumulation_steps", "effective_batch_size",
            "alpha", "beta", "gamma", "delta", "temperature", "skew_alpha",
            "layer_mapping_strategy", "layer_map_used",
            "loss_history", "epoch_times_seconds",
            "total_time_seconds", "avg_time_per_epoch"
        }
        self.assertEqual(required_keys, set(stats.keys()))

    def test_loss_history_has_five_keys(self):
        _, stats = distill_model(
            self.student, self.teacher, self.dataloader,
            epochs=1, show_progress=False, return_stats=True
        )
        self.assertEqual(
            set(stats["loss_history"].keys()),
            {"total", "task", "logits", "trajectory", "derivative"}
        )

    def test_layer_map_is_none_when_features_off(self):
        _, stats = distill_model(
            self.student, self.teacher, self.dataloader,
            gamma=0.0, delta=0.0,
            epochs=1, show_progress=False, return_stats=True
        )
        self.assertIsNone(stats["layer_map_used"])
        self.assertIsNone(stats["layer_mapping_strategy"])

    def test_epoch_times_length_matches_epochs(self):
        n_epochs = 2
        _, stats = distill_model(
            self.student, self.teacher, self.dataloader,
            epochs=n_epochs, show_progress=False, return_stats=True
        )
        self.assertEqual(len(stats["epoch_times_seconds"]), n_epochs)
        self.assertEqual(len(stats["loss_history"]["total"]), n_epochs)


if __name__ == "__main__":
    unittest.main()
```

---

## Acceptance Criteria

- [ ] `optipfair/distillation/trainer.py` exists with `distill_model()`
- [ ] `return_stats=False` returns only the trained model
- [ ] `return_stats=True` returns `(trained_model, stats_dict)` with the exact structure specified
- [ ] `layer_map_used` and `layer_mapping_strategy` are `None` in stats when `gamma=0` and `delta=0`
- [ ] `ValueError` is raised for: all weights zero, same model object, invalid strategy
- [ ] Device is inferred from student model, no global variable required
- [ ] All tests in `tests/test_distillation_trainer.py` pass
